core/nightshadev2/src/nightshade.rs

```diff
@@ -0,0 +1,370 @@
+    }
+}
+
+impl PartialOrd for State {
+    fn partial_cmp(&self, other: &State) -> Option<Ordering> {
+        self.bare_state.partial_cmp(&other.bare_state)
+    }
+}
+
+impl Ord for State {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.bare_state.cmp(&other.bare_state)
+    }
+}
+
+
+fn update_confidence1(state0: &mut State, state1: &State) {
+    // When this function is called it MUST hold that state0 >= state1. i.e.
+    // assert_eq!(state1 > state0, false);
+
+    if state0.endorses() != state1.endorses() {
+        if state1.bare_state.confidence0 > state0.bare_state.confidence1 {
+            state0.bare_state.confidence1 = state1.bare_state.confidence0;
+            state0.proof1 = state1.proof0.clone();
+        }
+    }
+    else{
+        if state1.bare_state.confidence1 > state0.bare_state.confidence1 {
+            state0.bare_state.confidence1 = state1.bare_state.confidence1;
+            state0.proof1 = state1.proof1.clone();
+        }
+    }
+}
+
+pub struct Nightshade {
+    owner_id: AuthorityId,
+    num_authorities: usize,
+    states: Vec<State>,
+    best_state_counter: usize,
+    seen_bare_states: HashSet<BareState>,
+    commited: bool,
+}
+
+impl Nightshade {
+    fn new(owner_id: AuthorityId, num_authorities: usize) -> Self {
+        let mut states = vec![];
+
+        for i in 0..num_authorities {
+            if i == owner_id{
+                states.push(State::new(i));
+            }
+            else{
+                states.push(State::empty());
+            }
+        }
+
+        Self {
+            owner_id,
+            num_authorities,
+            states,
+            best_state_counter: 1,
+            seen_bare_states: HashSet::new(),
+            commited: false,
+        }
+    }
+
+    fn state(&self) -> State {
+        self.states[self.owner_id].clone()
+    }
+
+    fn update_state(&mut self, authority_id: AuthorityId, state: State) -> NSResult {
+        if self.commited {
+            return NSResult::Updated(None);
+        }
+
+        // Verify this BareState only if it has not been successfully verified previously
+        if !self.seen_bare_states.contains(&state.bare_state) {
+            if state.verify() {
+                self.seen_bare_states.insert(state.bare_state.clone());
+            } else {
+                return NSResult::Error("Not valid state".to_string());
+            }
+        }
+
+        if state.bare_state > self.states[authority_id].bare_state {
+            self.states[authority_id] = state.clone();
+
+            // We always take the best state seen so far
+            if state > self.states[self.owner_id] {
+                self.best_state_counter = 1;
+
+                let mut tmp_state = state.clone();
+                update_confidence1(&mut tmp_state, &self.states[self.owner_id]);
+
+                self.states[self.owner_id] = tmp_state;
+            }
+
+            if state == self.states[self.owner_id] {
```

Comment:
we should compare BareState

Reply:
[Fixed](https://github.com/nearprotocol/nearcore/pull/549/commits/67b71d2feaf245696c38cc5b027521d22e87811c)